File: go/fixture/go-mini/internal/placement/picker.go
// Package placement picks the nodes that hold a snapshot's replicas.
package placement

import (
	"fmt"
	"log"
)

// Node is a storage node.
type Node struct {
	ID       string
	Zone     string
	Capacity int
}

// Placer is a placer.
type Placer struct {
	logger *log.Logger
}

// NewPlacer creates a new Placer.
func NewPlacer() *Placer {
	return &Placer{logger: log.Default()}
}

// Pick picks a primary node in zone and a secondary node outside it.
func (p *Placer) Pick(nodes []Node, zone string) (Node, Node, error) {
	primary, secondary, err := p.pickReplicas(nodes, zone)
	if err != nil {
		p.logger.Printf("placement: %v", err)
		return Node{}, Node{}, err
	}
	return primary, secondary, nil
}

func (p *Placer) pickReplicas(nodes []Node, zone string) (Node, Node, error) {
	var primary, secondary Node
	primaryFound, secondaryFound := false, false
	for _, n := range nodes {
		if n.Zone == "" || n.Capacity <= 0 {
			continue
		}
		if n.Zone == zone {
			if primaryFound {
				continue
			}
			primary, primaryFound = n, true
			continue
		}
		if secondaryFound {
			continue
		}
		secondary, secondaryFound = n, true
	}
	if !primaryFound || !secondaryFound {
		return Node{}, Node{}, fmt.Errorf("no placement for zone %q", zone)
	}
	return primary, secondary, nil
}

File: py/fixture/py-mini/internal/placement/picker.py
"""Picks the nodes that hold a snapshot's replicas."""

import logging
from dataclasses import dataclass


@dataclass
class Node:
    """A storage node."""

    id: str
    zone: str
    capacity: int


class Placer:
    """Placer is a placer."""

    def __init__(self) -> None:
        """Create a new Placer."""
        self.logger = logging.getLogger("placement")

    def pick(self, nodes: list[Node], zone: str) -> tuple[Node | None, Node | None, str | None]:
        """Pick a primary node in zone and a secondary node outside it."""
        primary, secondary, err = self._pick_replicas(nodes, zone)
        if err is not None:
            self.logger.info("placement: %s", err)
            return None, None, err
        return primary, secondary, None

    def _pick_replicas(
        self, nodes: list[Node], zone: str
    ) -> tuple[Node | None, Node | None, str | None]:
        primary: Node | None = None
        secondary: Node | None = None
        primary_found = False
        secondary_found = False
        for n in nodes:
            if not n.zone or n.capacity <= 0:
                continue
            if n.zone == zone:
                if primary_found:
                    continue
                primary = n
                primary_found = True
                continue
            if secondary_found:
                continue
            secondary = n
            secondary_found = True
        if not primary_found or not secondary_found:
            return None, None, f'no placement for zone "{zone}"'
        return primary, secondary, None
